Remove inline comments during line preprocessing

IniasP opened the pull request:
Looks like the current implementation only removes lines that start with "#" to filter comments.
As described in issue #160, inline comments are missed. This should resolve issue #160.

IniasP committed:
Match quoted return type hint

This resolves issue #147

IniasP committed:
Remove inline comments during line preprocessing

slmg commented on src/parse/utilities.ts:
```diff
@@ -15,7 +15,7 @@ export function getIndentation(line: string): string {
 export function preprocessLines(lines: string[]): string[] {
     return lines
-        .map(line => line.trim())
+        .map(line => line.split("#")[0].trim())
         .filter((line) => !line.startsWith("#"));
```
```suggestion
        .filter((line) => !line.startsWith("#"))
        .map(line => line.split("#")[0].trim());
```

Otherwise the condition `startsWith("#")` will never be true.

IniasP replied:
Oof, I totally missed that, good catch!

IniasP committed:
Correct line preprocessing order of operations

Co-authored-by: slmg <38896727+slmg@users.noreply.github.com>

NilsJPWerner commented:
Thank you for opening this PR! I will check it later this week.

IniasP committed:
Add test case for inline comment removal in preprocessing

Also fix the handling of lines like "    # comment" by trimming before the startsWith("#") check.
We can't just map once to the trimmed version, since the filter needs the trim before its check and the map can only trim after splitting on "#".

IniasP commented:
Good thing you asked for a test case, turns out we missed something (see commit message on 100adc3). Should be good now.

NilsJPWerner commented:
Sorry for the massive delay on responding to this. Does this handle strings that have the pound sign in them? e.g. `x = "# abc"`

NilsJPWerner commented on src/parse/utilities.ts:
```diff
@@ -15,8 +15,8 @@ export function getIndentation(line: string): string {
     return lines
-        .map(line => line.trim())
-        .filter((line) => !line.startsWith("#"));
+        .filter((line) => !line.trim().startsWith("#"))
```
We can't do this unfortunately because it will cut off strings with # in them. See previous discussion here: https://github.com/NilsJPWerner/autoDocstring/pull/144

NilsJPWerner commented:
So we could possibly limit this to only trim comments in the docstring definition since it is less likely to have a string with a # in it than in the body.

IniasP commented:
I don't think I realized at the time that hash symbols in strings are possible. I agree with the discussion at #144 that this will be impossible to do exactly right with regex, because it's fundamentally outside of the scope of regular languages.
Also, this is really low impact so maybe this should just be a wontfix thing.
Maybe there should be a test case addressing this specifically, because I did get all tests to pass.
